### PR TITLE
splitter: ingest+split logs from CWLogs subscription

### PR DESCRIPTION
**Pull request by nathanleiby**
Using this to test out new functionality. We may end up moving this to a separate service or lambda, to split logs before they get to the consumer.

### CODE REVIEW

**Review by xavi- — APPROVED** [review]
Lgtm.  Is there an example of how'd it be used by a consumer?

**Comment by nathanleiby** [comment]
@xavi- not yet, planning on testing today and making an example

**Comment by nathanleiby** [comment]
@xavi- Hey! Would you mind taking another pass on this one (when time permits), given all the changes since you original review?

In its current state, it should be

- backwards compatible
- able to handle a CWLogs subscription being added to a Kinesis stream

**Rollout Plan:**

I plan to deploy in Dev and monitor there for a bit first to make sure everything looks good. Then I'll enable a CW Logs subscription to the Dev kinesis stream, and verify that workflow logs are parsed correctly and arrive in Kibana / Archive.

After this, I'll deploy in prod.

**Comment by nathanleiby** [comment]
@xavi- One other note: this hasn't yet solved the problem of doing better checkpointing on Batch messages. Currently, if any message from the CW Logs record is checkpointed, it will assume the record was successful.

https://github.com/Clever/kinesis-to-firehose/blob/c7fc93a7b528776d68bb2d0ce8da44d1d130e36f/writer/firehose_writer.go#L160-L161

**Comment by nathanleiby** [comment]
I'm going to think on the right solution for that one, but I don't think it blocks deploying this in dev and being sure it works correctly.

**Comment by nathanleiby** [comment]
@xavi- there's one other workaround here that @mohit and I discussed offline. Given the way subscriptions work, it's hard to target dev logs to dev stream and prod logs to prod stream, so here's the workaround we came up with: https://github.com/Clever/kinesis-to-firehose/pull/21/commits/c7fc93a7b528776d68bb2d0ce8da44d1d130e36f

File: writer/firehose_writer.go [hunk 1]
```diff
@@ -136,14 +137,45 @@ func (f *FirehoseWriter) ProcessRecords(records []kcl.Record) error {
...
+			err := f.parseMessageAndPrepareToSend(m, record.SequenceNumber, record.SubSequenceNumber)
+			if err != nil {
+				lastErr = err
+			}
+		}
+		return lastErr
+	}
+
+	// Process a single message, from KPL
+	return f.parseMessageAndPrepareToSend(data, record.SequenceNumber, record.SubSequenceNumber)
```
Comment (xavi-) [inline writer/firehose_writer.go]: Nit. Perhaps reverse conditional here to:

```golang
if !splitter.IsGzipped(data) { // Process single message...
    // ...
} else { // Process messages from CWLogs
    // ...
    // ...
}
```

File: splitter/splitter.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+// IsGzipped returns whether or not a string is Gzipped (determined by looking for a Gzip byte prefix)
+func IsGzipped(s string) bool {
+	return strings.HasPrefix(s, "\x1f\x8b")
+}
+
+// GetMessagesFromGzippedInput takes a gzipped record from a CWLogs Subscription and splits it into a slice of messages.
+func GetMessagesFromGzippedInput(input string, prodEnv bool) ([]string, error) {
+	unpacked, err := unpack(input)
+	if err != nil {
+		fmt.Println("ERROR: " + err.Error())
```
Comment (xavi-) [inline splitter/splitter.go]: Debug statement?

**Review by xavi- — APPROVED** [review]

**Pull request merged**